subprojects/docs/src/docs/userguide/developing-plugins/writing_in_kotlin.adoc

```diff
@@ -0,0 +1,329 @@
+= Writing Gradle plugins in Kotlin
+
+In addition to <<kotlin_dsl.adoc, writing build scripts>>, Gradle also supports writing plugins in https://www.jetbrains.com/opensource/kotlin/[Kotlin].
+
+There are multiple options to write your Kotlin plugins as well as multiple compatibility considerations. This page is non-exhaustive list of dependencies and compiler configurations and their impacts in terms of compatibility and usability.
+
+It also describes the different development plugins to help you set up these different options.
+
+
+[[sec:dependencies]]
+== Different dependencies
+
+Like other JVM projects, Gradle plugins are jar files with dependencies. Unlike other JVM projects, Gradle provides fixed versions of some dependencies at runtime. Those dependencies are `compileOnly` dependencies and need to be compatible with the version of Gradle your plugin is running in.
+
+If you are writing convention plugins, this is usually not an issue. But if you distribute your plugins, you should pay attention to their compatibility.
+
+[[sec:gradle_api]]
+=== The Gradle API
+
+All plugins use the https://docs.gradle.org/current/javadoc/index.html[Gradle Core API] to configure a Gradle build. The Gradle API must be added as a build compile-time dependency using the helper function `gradleApi()`
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly(gradleApi())
+}
+----
+
+TIP: `gradleApi()` must be a `compileOnly` dependency, because the correct Gradle API will be provided by Gradle at runtime.
```

Comment:
> The `java-gradle-plugin` plugin automatically adds `gradleApi()` to `api`

I've always wondered the reason behind that actually. Is there a specific reason or is it just historical? Doesn't feel right to pull `gradleApi()` in the runtime classpath (even though as you state later those dependencies aren't propagated so it's basically the same result but still feels wrong)

> Given it's automatically added by `java-gradle-plugin` and that plugin should be used when developing the Gradle plugin, there is no need to declare a dependency on the Gradle API.

Agreed but I still think it's important to be aware what `gradleApi()` is? Maybe change the wording to something like this?

```markdown
TIP: `gradleApi()` is added to the `api` configuration by the `java-gradle-plugin` so you don't need to add it manually. If you do, it should be a `compileOnly` dependency, because the correct Gradle API will be provided by Gradle at runtime.
```

> Gradle doesn't provide any guidelines on support libraries for plugins. [...] supporting libraries must be published in Maven Central, which happens to be mirrored through the plugin portal so the plugin resolution can find them.

Agree this is weird and the main reason why I publish my plugins to both portal and Maven Central. 

All in all, looks like we need to move these Gradle APIs considerations to the [Implementing Plugins](https://docs.gradle.org/current/userguide/implementing_gradle_plugins.html) page. And then maybe link them from the Kotlin specific one? Let's wait a bit for Gradle's team feedback and if they're OK with it we can update this PR?

Reply:
> TIP: `gradleApi()` is added to the `api` configuration by the `java-gradle-plugin` so you don't need to add it manually

If `gradleApi()` is added by `java-gradle-plugin`, there's no need to adding it anyway.